src/Meshes/buffer.ts

```diff
@@ -1,6 +1,159 @@
+     * @param engine the engine
+     * @param data the data to use for this buffer
+     * @param updatable whether the data is updatable
+     * @param postponeInternalCreation whether to postpone creating the internal WebGL buffer (optional)
+     */
+    constructor (engine: ThinEngine, data: IndicesArray, updatable = false, postponeInternalCreation = false) {
+        this._engine = engine;
+        this._updatable = updatable;
+        this._data = data;
+        this._buffer = null;
+        if (!postponeInternalCreation) {
+            this.create();
+        }
+    }
+
+    /**
+     * Create the GPU memory for the IndexBuffer. Will have no effect if it was already created.
+     */
+    public create() {
+        if (this._buffer) {
+            return;
+        }
+        if (!this._buffer) {
+            this._buffer = this._engine.createIndexBuffer(this._data, this._updatable);
+        }
+        if (!this._restore_context_observer) {
+            this._restore_context_observer = this._engine.onContextRestoredObservable.add(this._rebuild);
+        }
+    }
+
+    /**
+     * Dispose the IndexBuffer.
+     * Will only do actual dispose if this was the last usage of the index buffer.
+     */
+    public dispose() {
+        if (this._buffer) {
+            if (this._engine._releaseBuffer(this._buffer)) {
+                this._buffer = null;
+                if (this._restore_context_observer) {
+                    this._engine.onContextRestoredObservable.remove(this._restore_context_observer);
+                }
+            }
+        }
+    }
+
+    /**
+     * Restore the IndexBuffer on the GPU. Used to restore buffer after a context lost.
+     */
+    private _rebuild() {
+        if (this._buffer === null) {
+            return;
+        }
+        // Make sure we retain theo riginal reference count of the buffer. Calling .create will reset it to 1.
+        const restore_references = this._buffer.references;
+        this._buffer = null;
+        this.create();
+        // The typescript compiler fails to realize that this.create() always recreates this._buffer, hence the ignore.
+        // @ts-ignore
+        this._buffer.references = restore_references;
+    }
+
+    /**
+     * Increase reference count to indicate that we have one more owner.
+     * This lets the outside call dispose() once more without deleting the actual buffer.
+     * @returns the same index buffer with a higher reference count
+     */
+    public makeNewReference(): IndexBuffer {
+        if (this._buffer) {
+            this._buffer.references++;
+        }
+        return this;
+    }
+
+    /**
+     * Update content of the buffer
+     * @param indices new indices to use
+     * @param offset optional offset in the target buffer
+     * @param gpuMemoryOnly specifies if the CPU memory should be updated
+     * @returns true if size changed
+     */
+    public update(indices: IndicesArray, offset?: number, gpuMemoryOnly = false): boolean {

```

Comment:
I'm wondering if we need a new class. Can we inherit from vertexbuffer and just change what is different? There is a lot of duplicated code

Other option: have a new parent class for vb and ib?

Reply:
I agree.

There is one more weird thing. I feel that the Buffer-class should be a parent-class to the VertexBuffer but instead the VertexBuffer HAS a Buffer. I can't think of a reason for this except that clones of VertexBuffers always should have the same buffer even if the internal DataBuffer is recreated. But there was no clone-function before...

My IndexBuffer class has a more direct relation to the DataBuffer (which is why the clone-function is called makeNewReference instead of clone).

Reply:
You can create a new VB and use a shared buffer (through constructor)